tests/utils/test_codegen.py

```diff
@@ -32,6 +32,106 @@ def test_compare_attributes(self):
         self.assertEqual(1, ClassUtils.INCLUDES_SOME)
         self.assertEqual(2, ClassUtils.INCLUDES_ALL)
 
+    @mock.patch.object(ClassUtils, "sanitize_sequential")
+    @mock.patch.object(ClassUtils, "sanitize_restrictions")
+    @mock.patch.object(ClassUtils, "sanitize_attribute")
+    def test_sanitize_attributes(
+        self,
+        mock_sanitize_attribute,
+        mock_sanitize_restrictions,
+        mock_sanitize_sequential,
+    ):
+
+        target = ClassFactory.elements(3)
+        inner = ClassFactory.elements(2)
+        target.inner.append(inner)
+
+        ClassUtils.sanitize_attributes(target)
+        ClassUtils.sanitize_attributes(ClassFactory.create())
+
+        mock_sanitize_attribute.assert_has_calls(
+            [
+                mock.call(target.attrs[0]),
+                mock.call(target.attrs[1]),
+                mock.call(target.attrs[2]),
+                mock.call(target.inner[0].attrs[0]),
+                mock.call(target.inner[0].attrs[1]),
+            ]
+        )
+        mock_sanitize_restrictions.assert_has_calls(
+            [
+                mock.call(target.attrs[0].restrictions),
+                mock.call(target.attrs[1].restrictions),
+                mock.call(target.attrs[2].restrictions),
+                mock.call(target.inner[0].attrs[0].restrictions),
+                mock.call(target.inner[0].attrs[1].restrictions),
+            ]
+        )
+        mock_sanitize_sequential.assert_has_calls(
+            [
+                mock.call(target.attrs, 0),
+                mock.call(target.attrs, 1),
+                mock.call(target.attrs, 2),
+                mock.call(target.inner[0].attrs, 0),
+                mock.call(target.inner[0].attrs, 1),
+            ]
+        )
+
+    def test_sanitize_attribute(self):
+        attr = AttrFactory.create(fixed=True)
+
+        ClassUtils.sanitize_attribute(attr)
+        self.assertTrue(attr.fixed)
+
+        attr.restrictions.max_occurs = 2
+        ClassUtils.sanitize_attribute(attr)
+        self.assertFalse(attr.fixed)
+
+    def test_sanitize_restrictions(self):
+        restrictions = [
+            Restrictions(min_occurs=0, max_occurs=0, required=True),
+            Restrictions(min_occurs=0, max_occurs=1, required=True),
+            Restrictions(min_occurs=1, max_occurs=1, required=False),
+            Restrictions(max_occurs=2, required=True),
+            Restrictions(min_occurs=2, max_occurs=2, required=True),
+        ]
+        expected = [
+            {},
+            {},
+            {"required": True},
+            {"max_occurs": 2, "min_occurs": 0},
+            {"max_occurs": 2, "min_occurs": 2},
+        ]
+
+        for idx, input in enumerate(restrictions):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Redefining built-in 'input'](https://app.codacy.com/manual/tefra/xsdata/pullRequest?prid=5244803)